app_dart/lib/src/request_handlers/refresh_chromebot_status.dart

```diff
@@ -79,19 +80,79 @@ class RefreshChromebotStatus extends ApiRequestHandler<Body> {
       DatastoreService datastore,
       Map<LuciBuilder, List<LuciTask>> luciTasks) async {
     for (String branch in branches) {
-      await for (FullTask task in datastore.queryRecentTasks(
-          taskName: builder.taskName, branch: branch)) {
-        for (LuciTask luciTask in luciTasks[builder]) {
+      final List<FullTask> tasks = await datastore
+          .queryRecentTasks(taskName: builder.taskName, branch: branch)
+          .toList();
+      final List<LuciTask> allLuciTasks = luciTasks[builder];
+      final Set<LuciTask> updatedLuciTasks = <LuciTask>{};
+
+      /// Scan both [tasks] and [luciTasks] from old to new, since [luciTasks] may
+      /// contain new re-run builds which are not in datastore yet. If matched, then
+      /// update accordingly.
+      for (int i = tasks.length - 1; i >= 0; i--) {
+        final FullTask task = tasks[i];
+        for (int j = allLuciTasks.length - 1; j >= 0; j--) {
+          final LuciTask luciTask = allLuciTasks[j];
           if (luciTask.commitSha == task.commit.sha &&
-              luciTask.ref == 'refs/heads/$branch') {
+              luciTask.ref == 'refs/heads/$branch' &&
+              (task.task.buildId == null ||
+                  task.task.buildId == luciTask.buildId) &&
+              !updatedLuciTasks.contains(luciTask)) {
```

Comment:
Looks like a very complex if to compare two objects. Do we need to override operator == in either luci or task to avoid this?

Reply:
I have put these conditions to a sub-function to make it look simpler.

Reply:
Moving it to a to a function is just moving the complexity to a different place. I'd suggest trying to remove the complexity instead.

Reply:
Inspired one you your comments above about using `Map`, I have including `sha` in the returned `luciTasks`. This naturally removes two conditions, together with the `branch`.